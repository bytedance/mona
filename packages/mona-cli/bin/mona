#! /usr/bin/env node

const mona = require('../dist').default;
mona();
