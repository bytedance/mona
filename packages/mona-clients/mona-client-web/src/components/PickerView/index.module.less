.pickerView {
  position: relative;
  width: 100%;
  z-index: 2;
  margin: 0.5em 0;

  &Multi {
    color: #1d2129;
    font-size: 0.5em;
  }

  &Multi,
  &Selection {
    display: flex;
    align-items: center;
    position: absolute;
    left: 0;
    top: 0;
    width: 100%;
    height: 100%;
  }

  &Column {
    flex: 1;
    height: 100%;
    overflow: hidden;
    text-align: center;

    &ItemWrap {
      box-sizing: content-box;
      will-change: transform;
    }
  }

  &Selection {
    will-change: transform;
    flex-direction: column;
    pointer-events: none;

    &MaskTop {
      width: 100%;
      flex: 1;
      background: rgba(255, 255, 255, 0.404);
    }

    &MaskBottom {
      width: 100%;
      flex: 1;
      background: rgba(255, 255, 255, 0.404);
    }

    &Bar {
      width: 100%;
      height: 1em;
      flex: 0 0 auto;
      top: 1px solid #e5e6eb;
      bottom: 1px solid #e5e6eb;
      border: 1px solid #e5e6eb;
    }
  }
}
