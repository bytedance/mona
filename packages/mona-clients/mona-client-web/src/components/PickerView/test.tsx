// import React, { useEffect, useState, useRef, forwardRef, Ref, useImperativeHandle } from 'react';
// import { cls } from '@arco-design/mobile-utils';
// import { ContextLayout } from '../context-provider';
// import Popup from '../popup';
// import PickerView, { PickerViewRef, ValueType, PickerData, PickerCellMovingStatus } from '../picker-view';
// import { PickerProps, DataType } from './type';
// import { useListenResize } from '../_helpers';

// export * from './type';
// export { MultiPicker, PickerCell, Cascader } from '../picker-view';

// export interface PickerRef {
//   /** 最外层元素 DOM */
//   dom: HTMLDivElement | null;
//   /** 每一列的滑动状态 */
//   getCellMovingStatus: () => PickerCellMovingStatus[];
//   /** 手动更新元素布局 */
//   updateLayout: () => void;
//   /** 获取所有列的值 */
//   getAllColumnValues: () => ValueType[];
//   /** 获取第 n 列的值 */
//   getColumnValue: (index: number) => ValueType;
// }

// const getInitialValue = (value, data: DataType, cascade) => {
//   if (value && value.length) {
//     return value;
//   }

//   const computedValue: ValueType[] = [];

//   if (!cascade) {
//     if (!(data[0] instanceof Array)) {
//       return data[0]?.value ? [data[0].value] : [];
//     }

//     (data as (ValueType | PickerData)[][]).map(column => {
//       computedValue.push(typeof column[0] === 'object' ? column[0].value : column[0]);
//     });
//   } else {
//     const cascadePickerData = data as unknown as PickerData[];
//     if (!cascadePickerData.length) {
//       return computedValue;
//     }

//     computedValue.push(cascadePickerData[0].value);

//     let traverse = cascadePickerData[0].children;

//     while (traverse) {
//       computedValue.push(traverse[0].value);

//       traverse = traverse[0].children;
//     }
//   }

//   return computedValue;
// };

// /**
//  * 选择器组件，形式是弹起的浮层。
//  * @type 数据录入
//  * @name 选择器
//  */
// const Picker = forwardRef((props: PickerProps, ref: Ref<PickerRef>) => {
//   const {
//     className,
//     itemStyle,
//     cascade = true,
//     cols = 3,
//     rows = 5,
//     data,
//     okText = '确定',
//     dismissText = '取消',
//     disabled = false,
//     clickable = true,
//     hideEmptyCols = false,
//     title = '',
//     visible = false,
//     value,
//     needBottomOffset = false,
//     onDismiss,
//     onOk,
//     onChange,
//     maskClosable = false,
//     onHide,
//     onColumnChange,
//     ...otherProps
//   } = props;

//   const [scrollValue, setScrollValue] = useState(getInitialValue(value, data, cascade));
//   const domRef = useRef<HTMLDivElement | null>(null);
//   const pickerViewRef = useRef<PickerViewRef>(null);

//   useImperativeHandle(ref, () => ({
//     dom: domRef.current,
//     getCellMovingStatus: () => pickerViewRef.current?.getCellMovingStatus() || [],
//     getAllColumnValues: () => pickerViewRef.current?.getAllColumnValues() || [],
//     getColumnValue: index => pickerViewRef.current?.getColumnValue(index) || '',
//     updateLayout: () => pickerViewRef.current?.updateLayout(),
//   }));

//   function handleDismiss() {
//     if (onDismiss) {
//       onDismiss();
//     }
//     if (onHide) {
//       onHide('dismiss');
//     }
//   }

//   const handleConfirm = () => {
//     const val = pickerViewRef.current?.getAllColumnValues() || scrollValue;
//     if (onOk) {
//       onOk(val);
//     }
//     if (onChange) {
//       onChange(val);
//     }
//     if (onHide) {
//       onHide('confirm');
//     }
//   };

//   useEffect(() => {
//     setScrollValue(value);
//   }, [value, setScrollValue]);

//   useListenResize(updateLayoutByVisible, [visible]);

//   // 每次visible从false变为true时需要重新设置scrollValue的值为当前value的值（初始值）
//   function updateLayoutByVisible() {
//     if (visible && pickerViewRef.current) {
//       pickerViewRef.current.updateLayout();
//     }
//   }

//   useEffect(() => {
//     updateLayoutByVisible();
//   }, [visible]);

//   return (
//     <ContextLayout>
//       {({ prefixCls }) => (
//         <>
//           <Popup
//             visible={visible}
//             className={className + ` ${prefixCls}-picker all-border-box`}
//             close={() => onHide?.('mask')}
//             direction="bottom"
//             maskClosable={maskClosable}
//             needBottomOffset={needBottomOffset}
//             {...otherProps}
//           >
//             <div className={`${prefixCls}-picker-wrap`} ref={domRef}>
//               <div className={`${prefixCls}-picker-header`}>
//                 <div className={`${prefixCls}-picker-header-btn left`} onClick={handleDismiss}>
//                   {dismissText}
//                 </div>
//                 <div className={`${prefixCls}-picker-header-title`}>{title}</div>
//                 <div className={`${prefixCls}-picker-header-btn right`} onClick={handleConfirm}>
//                   {okText}
//                 </div>
//               </div>
//               <PickerView
//                 ref={pickerViewRef}
//                 data={data}
//                 cascade={cascade}
//                 cols={cols}
//                 rows={rows}
//                 disabled={disabled}
//                 value={getInitialValue(value, data, cascade)}
//                 onColumnChange={onColumnChange}
//                 itemStyle={itemStyle}
//                 clickable={clickable}
//                 hideEmptyCols={hideEmptyCols}
//               />
//             </div>
//           </Popup>
//         </>
//       )}
//     </ContextLayout>
//   );
// });

// export default Picker;
