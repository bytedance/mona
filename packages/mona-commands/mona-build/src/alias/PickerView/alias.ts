import { bubbleEventsAlias, basePropAlias, propAliasMap } from '../prop';
import { PickerViewAlias as Alias } from '../type';

const alias: Alias = {
  ...basePropAlias,
  ...bubbleEventsAlias,
  value: propAliasMap.value,
  indicatorStyle: propAliasMap.indicatorStyle,
  maskStyle: propAliasMap.maskStyle,
  onChange: propAliasMap.onChange,
};

export default alias;
