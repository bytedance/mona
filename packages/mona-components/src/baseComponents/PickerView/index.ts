import React from 'react';
import createComponent from '../../createComponent';
import { BaseProps, EventHandler } from '../base';

export interface PickerViewProps extends BaseProps {
  value: number[];
  indicatorStyle?: string;
  maskStyle?: string;
  onChange?: EventHandler
}

export const PickerView: React.ComponentType<PickerViewProps> = createComponent<PickerViewProps>('picker-view')

