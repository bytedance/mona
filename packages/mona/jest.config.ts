import baseConfig from '../../jestconfig.base';

module.exports = baseConfig;
