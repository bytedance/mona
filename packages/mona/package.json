{
  "name": "@bytedance/mona",
  "version": "0.2.0-alpha.9",
  "description": "Merchant app's developing and building tools",
  "main": "./dist/index",
  "scripts": {
    "build": "yarn clean && ttsc -b",
    "start": "ttsc -b -w",
    "clean": "rm -rf dist",
    "test": "jest"
  },
  "files": [
    "dist"
  ],
  "author": "zhangxiaowei.chris@bytedance.com",
  "license": "MIT",
  "publishConfig": {
    "access": "public",
    "registry": "https://registry.npmjs.org"
  },
  "devDependencies": {
    "copy-webpack-plugin": "^10.2.0",
    "react": "^17.0.2",
    "webpack": "^5.58.2",
    "webpack-chain": "^6.5.1"
  },
  "peerDependencies": {
    "react": "^17.0.2"
  },
  "gitHead": "cc5aa5283af595fd7d8a7b4921f54e2797504edd"
}
