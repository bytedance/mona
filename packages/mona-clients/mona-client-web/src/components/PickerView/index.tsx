import React, { useEffect, useMemo, useRef, useState } from 'react';
import { PickerViewProps } from '@bytedance/mona';
import { isElement } from 'react-is';
import { useHandlers } from '../hooks';

import styles from './index.module.less';
import { genEvent } from '../utils';

const PickerView: React.FC<PickerViewProps> = ({
  children,
  value,
  indicatorStyle,
  maskStyle,
  onChange,
  ...restProps
}) => {
  const [itemHeight, setItemHeight] = useState(0);
  const barRef = useRef<HTMLDivElement | null>(null);
  const wrapperRef = useRef<HTMLDivElement>(null);
  const [wrapperHeight, setWrapperHeight] = useState(0);
  const { handleClassName, ...handleProps } = useHandlers(restProps);
  const innerValueRef = useRef(value ?? []);
  useEffect(() => {
    if (wrapperRef.current) {
      setWrapperHeight(wrapperRef.current.offsetHeight);
    }
    if (barRef.current) {
      setItemHeight(barRef.current.clientHeight);
    }
  }, []);

  function _onValueChange(index: number, nValue: any) {
    const newValue = innerValueRef.current?.concat() ?? [];
    newValue[index] = nValue;
    innerValueRef.current = newValue;
    onChange?.(genEvent({ type: 'change', detail: { value: newValue } }));
  }

  return (
    <div className={styles.pickerView} style={{ height: `${itemHeight * 5}px` }} ref={wrapperRef} {...handleProps}>
      <div className={styles.pickerViewMulti} style={{ lineHeight: `${itemHeight}px` }}>
        {React.Children.map(children, (item, idx) => {
          //@ts-ignore
          if (isElement(item) && item.type.name === 'PickerViewColumn') {
            return React.cloneElement(item, {
              itemHeight,
              wrapperHeight,
              selectedValue: value[idx],
              onValueChange: _onValueChange.bind(null, idx),
            });
          }
        })?.filter(item => item !== null && item !== undefined)}
      </div>

      <div className={styles.pickerViewSelection} style={maskStyle}>
        <div className={styles.pickerViewSelectionMaskTop} />
        <div ref={barRef} className={styles.pickerViewSelectionBar} style={indicatorStyle} />
        <div className={styles.pickerViewSelectionMaskTop} />
      </div>
    </div>
  );
};

export default PickerView;
