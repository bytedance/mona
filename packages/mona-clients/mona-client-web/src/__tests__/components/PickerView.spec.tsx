import React from 'react';
import { mount } from 'enzyme';
import PickerView from '../../components/PickerView';
import mountTest from '../../../../../tests/shared/mountTest';
import handlerTest from '../../../../../tests/shared/handlerTest';
import { PickerViewProps } from '@bytedance/mona';

describe('web component: PickerView', () => {
  mountTest(PickerView)
  handlerTest({ Component: PickerView, firstTag: 'div' });

  it('should render correctly', () => {
    expect(mount(<PickerView value={[]} />).render()).toMatchSnapshot();
  })
  
  it('should have right properties', () => {
    const props: PickerViewProps = {
      value: [],
      className: 'class-name',
      style: { color: 'red' },
    }
    const wrapper = mount(<PickerView {...props}>hello</PickerView>);
    expect(wrapper.render()).toMatchSnapshot();
  })
})
