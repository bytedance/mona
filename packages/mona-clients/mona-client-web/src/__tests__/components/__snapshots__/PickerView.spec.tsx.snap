// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`web component: PickerView should have right properties 1`] = `
LoadedCheerio {
  "0": <div
    class="pickerView"
    style="color: red;"
  >
    <div
      class="pickerViewMulti"
      style="line-height: 0px;"
    />
    <div
      class="pickerViewSelection"
    >
      <div
        class="pickerViewSelectionMaskTop"
      />
      <div
        class="pickerViewSelectionBar"
      />
      <div
        class="pickerViewSelectionMaskTop"
      />
    </div>
  </div>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`web component: PickerView should render correctly 1`] = `
LoadedCheerio {
  "0": <div
    class="pickerView"
    style="height: 0px;"
  >
    <div
      class="pickerViewMulti"
      style="line-height: 0px;"
    />
    <div
      class="pickerViewSelection"
    >
      <div
        class="pickerViewSelectionMaskTop"
      />
      <div
        class="pickerViewSelectionBar"
      />
      <div
        class="pickerViewSelectionMaskTop"
      />
    </div>
  </div>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;
